samples/socket-api/wasm-src/socket_opts.c

```diff
@@ -260,3 +273,162 @@ main(int argc, char *argv[])
+    addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    socket_fd = guard(socket(AF_INET, SOCK_STREAM, 0), "Create socket failed");
+    guard(set_and_get_bool_opt(socket_fd, SOL_SOCKET, SO_REUSEADDR, 1),
+          "Failed to set REUSEADDR");
+    guard(bind(socket_fd, (struct sockaddr *)&addr, addrlen),
+          "Bind socket failed");
+    guard(listen(socket_fd, 1), "Listen failed");
+
+    printf("Wait for client to connect\n");
+    client_socket_fd = guard(
+        accept(socket_fd, (struct sockaddr *)&addr, (socklen_t *)&addrlen),
+        "Accept failed");
+
+    printf("Client connected, sleeping for 10s\n");
+    sleep(10);
+
+    printf("Shuting down\n");
+    shutdown(client_socket_fd, SHUT_RDWR);
+    shutdown(socket_fd, SHUT_RDWR);
+    return EXIT_SUCCESS;
+}
+
+int 
+test_multicast_client() {
+    struct sockaddr_in localSock;
+    struct ip_mreq group;
+    int sd;
+    int datalen;
+    char databuf[1024];
+    sd = guard(socket(AF_INET, SOCK_DGRAM, 0), "Failed opening socket");
+    guard(set_and_get_bool_opt(sd, SOL_SOCKET, SO_REUSEADDR, 1), "Failed setting SO_REUSEADDR");
+
+    localSock.sin_family = AF_INET;
+    localSock.sin_port = htons(4321);
+    localSock.sin_addr.s_addr = INADDR_ANY;
+
+    guard(bind(sd, (struct sockaddr*)&localSock, sizeof(localSock)), "Failed binding socket");
+
+    group.imr_multiaddr.s_addr = MULTICAST_ADDR;
+    group.imr_interface.s_addr = htonl(INADDR_ANY);
+
+    guard(setsockopt(sd, IPPROTO_IP, IP_ADD_MEMBERSHIP, (char *)&group, sizeof(group)), "Failed adding multicast group");
+
+    datalen = sizeof(databuf);
+    int result = read(sd, databuf, datalen);
+
+    OPTION_ASSERT((result < 0), 0, "read response");
+
+    printf("Reading datagram message...OK.\n");
+    printf("The message from multicast server is: \"%s\"\n", databuf);
+
+    return EXIT_SUCCESS;
+}
+
+int 
+test_multicast_server() {
+    struct in_addr localInterface;
+    struct sockaddr_in groupSock;
+    int sd;
+    char * databuf = "Test message";
+    int datalen = strlen(databuf);
+    sd = guard(socket(AF_INET, SOCK_DGRAM, 0), "Failed to open socket");
+
+    groupSock.sin_family = AF_INET;
+    groupSock.sin_addr.s_addr = MULTICAST_ADDR;
+    groupSock.sin_port = htons(4321);
+    localInterface.s_addr = htons(INADDR_ANY);
+
+    guard(setsockopt(sd, IPPROTO_IP, IP_MULTICAST_IF, (char *)&localInterface, sizeof(localInterface)), "Failed setting local interface");
+    guard(sendto(sd, databuf, datalen, 0, (struct sockaddr*)&groupSock, sizeof(groupSock)), "Failed sending datagram");
+
+    return EXIT_SUCCESS;
+}
+
+#define RUN_WITH_ARG(ARG, NAME, FUNC) \
+    if (strcmp(ARG, NAME) == 0) {     \
+        guard(FUNC(), "Failed");      \
+    }
+
+int
+main(int argc, char *argv[])
+{
+    if (argc < 2) {
+        printf("Usage: %s <setget|timeout_client|timeout_server|multicast_client|multicast_server>\n", argv[0]);
+        return EXIT_FAILURE;
+    }
+
+    RUN_WITH_ARG(argv[1], "setget", test_set_and_get_opts)
```

Comment:
I wonder if there's any benefit of having all those programs as a single application or maybe we could have a multiple standalone tiny examples? I think from reader's perspective it might be better to focus on one functionality rather than having a huge file with various examples.

Reply:
This is what I was thinking initially, but I felt it was odd to have a new file for a server that doesn't do anything and to label it as a sockopts example. Thinking back, I think it's better to have the smaller examples so will revise